sdk/test-utils/recorder/src/baseRecorder.ts

```diff
@@ -0,0 +1,414 @@
+  private matchRequest(recording: any, request: any): boolean {
+    // Every parameter in the recording must be present and have the same value in the request
+    for (const param in recording.query) {
+      if (recording.query[param] !== request.query[param]) {
+        return false;
+      }
+    }
+
+    // There shouldn't be parameters in the request that are not present in the recording (except for SAS Query Parameters and "_")
+    for (const param in request.query) {
+      if (
+        recording.query[param] === undefined &&
+        !this.sasQueryParameters.includes(param) &&
+        param !== "_"
+      ) {
+        return false;
+      }
+    }
+
+    return (
+      recording.method === request.method &&
+      recording.url === request.url &&
+      recording.requestBody === request.requestBody
+    );
+  }
+
+  // When recording, we want to hit the server and intercept requests/responses
+  // Nise does not allow us to intercept requests if they're sent to the server, so we need to override its behavior
+  public record(): void {
+    const self = this;
+    const xhr = nise.fakeXhr.useFakeXMLHttpRequest();
+
+    // The following filter allows every request to be sent to the server without being mocked
+    xhr.useFilters = true;
+    xhr.addFilter(() => true);
+
+    // 'onCreate' function is called when a new fake XMLHttpRequest object (req) is created
+    // Our intent is to override the request's 'onreadystatechange' function so we can create a recording once the response is ready
+    // We can only override 'onreadystatechange' AFTER the 'send' function is called because we need to make sure our implementation won't be overriden by the client
+    // But we can only override 'send' AFTER the 'open' function is called because the filter we set above makes Nise override it in 'open' body
+    xhr.onCreate = function(req: any) {
+      // We'll override the 'open' function, so we need to store a handle to its original implementation
+      const reqOpen = req.open;
+      req.open = function() {
+        // Here we are calling the original 'open' function to make sure everything is set up correctly (HTTP method, url, filters)
+        reqOpen.apply(req, arguments);
+
+        // We'll override the 'send' function, so we need to store a handle to its original implementation
+        // We can already override it because we know 'open' has already been called
+        const reqSend = req.send;
+        req.send = function(data: any) {
+          // We'll override the 'onreadystatechange' function, so we need to store a handle to its original implementation
+          // Now we can finally override 'onreadystatechange' because 'send' has already been called
+          const reqStateChange = req.onreadystatechange;
+          req.onreadystatechange = function() {
+            // Record the request once the response is obtained
+            if (req.readyState === 4) {
```

Comment:
Can you say more about what readyState is, and what in particular 4 is? Could this be an enum instead?

Reply:
Added some comments stating what `readyState` refers to.

Reply:
I mean more the hard-coded 4. Why is it 4 and not 3 or 5?

Reply:
More info - https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState
![image](https://user-images.githubusercontent.com/10452642/63553698-1da41b00-c4f0-11e9-8ea8-1ed622d4f442.png)




Added this comment - `readyState = 4 refers to the completion of the operation.`. Should I talk more about it?

Reply:
Ah okay, so there isn't an official enumeration, just the number. It's fine to leave it as is, but you may want to link to the documentation for people who try to read the code later